---
title: "Using proteasy to Retrieve Protease Data"
author: "Martin RydÃ©n"
package: proteasy
output:
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{Using proteasy to retrieve proteases acting on cleavage sites}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteKeywords{Proteomics, Bioinformatics, Cleavage, Polypeptides, MEROPS, Proteases}
  %\VignetteEncoding{UTF-8}
  %\VignettePackage{proteasy}
---

```{r biocstyle, echo = FALSE, results = "asis", message = FALSE}
library(BiocStyle)
BiocStyle::markdown()
```

# Introduction

This package utilizes data derived from the MEROPS database (https://www.ebi.ac.uk/merops/). Similarly to existing tools such as TopFind (https://topfind.clip.msl.ubc.ca/), `r BiocStyle::Biocpkg("proteasy")` exists for the purpose of retrieving data on proteolytic enzymes by mapping peptide termini to known sites where a protease cleaves. Unlike the `r BiocStyle::Biocpkg("cleaver")` package, which allows for in-silico cleavage of amino acid sequences by specifying an enzyme, the functions in `r BiocStyle::Biocpkg("proteasy")` relies only on experimentally derived data to find proteases annotated to cleavage sites.

This document illustrates the functionality of proteasy through some use cases.

# Installation

The package should be installed using as described below:

```{r installPackage, eval = FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("proteasy")

```

The package can then be loaded with:

```{r loadPackage}

library("proteasy")

```

# Usage

## Find known cleavages for a protein

```{r Search substrate, eval = TRUE}

# Kininogen-1
searchSubstrate(protein = "P01042", summarize = TRUE)

```


## Find known substrates a protease cleaves

```{r Show protease, eval = TRUE}

# MMP12
searchProtease(protein = "P39900", summarize = TRUE)

```

## Find possible proteases for cleaved substrates

```{r Find protease, eval = TRUE}

protein <- c("P02671", "P02671", "P68871", "P01011")
peptide <- c("FEEVSGNVSPGTR", "FVSETESR", "LLVVYPW", "ITLLSAL")

# If we do not specify start position (start_pos) and end position (end_pos),
# the function will automatically assign these values by matching the provided
# peptide sequence against the full-length protein sequence
res <- findProtease(protein = protein, peptide = peptide, organism = "Homo sapiens")

# The resulting S4 object can be inspected in three ways;

# 1. Display sequence data for the provided input:

substrates(res)

# 2. Show which known proteases may have cleaved this protein:

proteases(res)

# 3. Display details of associated proteolytic events:

cleavages(res)


```

## Look up a protease on MEROPS

```{r Browse protease, eval = FALSE}

browseProtease("P07339", keytype = "UniprotID") # (opens web browser)

```


# Session Information

```{r sessioninfo, echo=FALSE}
sessionInfo()
```
